src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +619,263 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoins descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ *
+ * Returns true if given node is recurring.
+ */
+static bool
+RecursivelyPlanRecurringTupleOuterJoins(Node *node, Query *query,
+										RecursivePlanningContext *recursivePlanningContext)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		return IsRTERefRecurring((RangeTblRef *) node, query);
+	}
+	else if (IsA(node, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) node;
+		ListCell *fromExprCell;
+
+		bool allRecurring = true;
+
+		/* search for join trees in each FROM element */
+		foreach(fromExprCell, fromExpr->fromlist)
+		{
+			Node *fromElement = (Node *) lfirst(fromExprCell);
+
+			if (!RecursivelyPlanRecurringTupleOuterJoins(fromElement, query,
+														 recursivePlanningContext))
+			{
+				allRecurring = false;
+			}
+		}
+
+		return allRecurring;
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) node;
+
+		Node *leftNode = joinExpr->larg;
+		Node *rightNode = joinExpr->rarg;
+
+		/*
+		 * There may be recursively plannable outer joins deeper in the join tree.
+		 *
+		 * We first handle the sub join trees and then the top level one since the
+		 * top level join expression might not require recursive planning after
+		 * handling the sub join trees.
+		 */
+		bool leftNodeRecurs = RecursivelyPlanRecurringTupleOuterJoins(leftNode, query,
+																	  recursivePlanningContext);
+		bool rightNodeRecurs = RecursivelyPlanRecurringTupleOuterJoins(rightNode, query,
+																	   recursivePlanningContext);
+		switch (joinExpr->jointype)
+		{
+			case JOIN_LEFT:
+			{
+				/* <recurring> left join <distributed> */
+				if (leftNodeRecurs && !rightNodeRecurs)
+				{
+					ereport(DEBUG1, (errmsg("recursively planning right side of "
+											"the left join since the outer side "
+											"is a recurring rel")));
+					RecursivelyPlanDistributedJoinNode(rightNode, query,
+													   recursivePlanningContext);
+				}
+
+				return leftNodeRecurs;
```

Comment:
Indeed, say for a LEFT JOIN, rhs being recurring or not doesn't contribute the recurring-ness of the join tree, we can say that a left join is recurring if the outer side recurring.

When the rhs is distributed but the lhs is recurring then the join would return a wrong result set, but is still recurring.

So let me add such a comment for left join and but simpler comments for the other join types.
```c
/*
 * A LEFT JOIN is recurring if the lhs is recurring.
 * Note that it doesn't matter whether the rhs is recurring
 * because if the rhs is recurring, then we should have already
 * convert rhs to a recurring rel too a few lines above. If we
 * couldn't do so for some reason, then this join would still be
 * a recurring one but would return incorrect results; hence we
 * would anyway throw an error for it later in the planner --via
 * DeferredErrorIfUnsupportedRecurringTuplesJoin.
 */
```